### ISSUE TITLE
Verbose message about missing dependencies

### ISSUE THREAD

**Issue by ad-m** [deanmalmgren/textract#130]
Hello,

I am would like tests textract tesseract integration. I installed it on VM, but I miss ```tesseract-ocr```. Only ```tesseract-lang-pol``` package installed.

So I got error message:
```
Traceback (most recent call last):
  File "/tmp/tmp.NitGbY38WB/bin/textract", line 32, in <module>
    main()
  File "/tmp/tmp.NitGbY38WB/bin/textract", line 25, in main
    output = process(**vars(args))
  File "/tmp/tmp.NitGbY38WB/local/lib/python2.7/site-packages/textract/parsers/__init__.py", line 59, in process
    return parser.process(filename, encoding, **kwargs)
  File "/tmp/tmp.NitGbY38WB/local/lib/python2.7/site-packages/textract/parsers/utils.py", line 45, in process
    byte_string = self.extract(filename, **kwargs)
  File "/tmp/tmp.NitGbY38WB/local/lib/python2.7/site-packages/textract/parsers/image.py", line 20, in extract
    stdout, _ = self.run(args)
  File "/tmp/tmp.NitGbY38WB/local/lib/python2.7/site-packages/textract/parsers/utils.py", line 83, in run
    stdout=subprocess.PIPE, stderr=subprocess.PIPE,
  File "/usr/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1327, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

It doesn't explain anything. I found missing dependencies, after update ```/tmp/tmp.NitGbY38WB/local/lib/python2.7/site-packages/textract/parsers/utils.py```, but the errors should mention missing file explicty.

Greetings,

**Comment by deanmalmgren** [comment]
Hmmmm... That's unexpected. textract usually displays [an error message if the command isn't installed](https://github.com/deanmalmgren/textract/blob/master/textract/exceptions.py#L69-L76). Can you share some more details about how to reproduce this error?

**Comment by ad-m** [comment]
```
(tmp.mRzWi4OSaS) vagrant@web1:~$ mktemp_virtualenv 
New python executable in /tmp/tmp.OkPrmL4Cv0/bin/python
Installing setuptools, pip, wheel...done.
(tmp.OkPrmL4Cv0) vagrant@web1:~$ pip install textract
Collecting textract
Collecting docx2txt (from textract)
Collecting EbookLib (from textract)
Collecting beautifulsoup4 (from textract)
  Using cached beautifulsoup4-4.5.1-py2-none-any.whl
Collecting SpeechRecognition>=3.1.0 (from textract)
Collecting chardet (from textract)
Collecting six (from textract)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting argcomplete (from textract)
  Using cached argcomplete-1.6.0-py2.py3-none-any.whl
Collecting python-pptx>=0.5.1 (from textract)
Collecting xlrd (from textract)
Collecting lxml (from EbookLib->textract)
  Using cached lxml-3.6.4-cp27-cp27mu-manylinux1_x86_64.whl
Collecting Pillow>=2.6.1 (from python-pptx>=0.5.1->textract)
  Using cached Pillow-3.4.2-cp27-cp27mu-manylinux1_x86_64.whl
Collecting XlsxWriter>=0.5.7 (from python-pptx>=0.5.1->textract)
  Using cached XlsxWriter-0.9.3-py2.py3-none-any.whl
Installing collected packages: docx2txt, lxml, six, EbookLib, beautifulsoup4, SpeechRecognition, chardet, argcomplete, Pillow, XlsxWriter, python-pptx, xlrd, textract
Successfully installed EbookLib-0.15 Pillow-3.4.2 SpeechRecognition-3.5.0 XlsxWriter-0.9.3 argcomplete-1.6.0 beautifulsoup4-4.5.1 chardet-2.3.0 docx2txt-0.6 lxml-3.6.4 python-pptx-0.6.1 six-1.10.0 textract-1.5.0 xlrd-1.0.0
(tmp.OkPrmL4Cv0) vagrant@web1:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.5 LTS
Release:	14.04
Codename:	trusty
(tmp.OkPrmL4Cv0) vagrant@web1:~$ cd tests/
(tmp.OkPrmL4Cv0) vagrant@web1:~/tests$ textract --method=tesseract -O language=pol OR-II-P.1431.123.2016.pdf
Traceback (most recent call last):
  File "/tmp/tmp.OkPrmL4Cv0/bin/textract", line 32, in <module>
    main()
  File "/tmp/tmp.OkPrmL4Cv0/bin/textract", line 25, in main
    output = process(**vars(args))
  File "/tmp/tmp.OkPrmL4Cv0/local/lib/python2.7/site-packages/textract/parsers/__init__.py", line 58, in process
    return parser.process(filename, encoding, **kwargs)
  File "/tmp/tmp.OkPrmL4Cv0/local/lib/python2.7/site-packages/textract/parsers/utils.py", line 45, in process
    byte_string = self.extract(filename, **kwargs)
  File "/tmp/tmp.OkPrmL4Cv0/local/lib/python2.7/site-packages/textract/parsers/pdf_parser.py", line 33, in extract
    return self.extract_tesseract(filename, **kwargs)
  File "/tmp/tmp.OkPrmL4Cv0/local/lib/python2.7/site-packages/textract/parsers/pdf_parser.py", line 61, in extract_tesseract
    page_content = TesseractParser().extract(page_path, **kwargs)
  File "/tmp/tmp.OkPrmL4Cv0/local/lib/python2.7/site-packages/textract/parsers/image.py", line 20, in extract
    stdout, _ = self.run(args)
  File "/tmp/tmp.OkPrmL4Cv0/local/lib/python2.7/site-packages/textract/parsers/utils.py", line 83, in run
    stdout=subprocess.PIPE, stderr=subprocess.PIPE,
  File "/usr/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1327, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

**Comment by webeng** [comment]
Same issue here. It doesn't work for any format...

**Comment by scubedoo187** [comment]
Same issue here. Does not work any format, path...

**Comment by webeng** [comment]
I fixed the problem by installing the required dependencies in the installation section: http://textract.readthedocs.io/en/stable/installation.html

The link in the README should link to the installation page as some of us just did pip install extract before installing the required dependencies before installing textract.

**Comment by deanmalmgren** [comment]
It seems like we need to have a general way of alerting users that dependencies are missing. Does anyone know how other packages handle this?

The issue is that I don't think we want to to require users to install all dependencies on their system at installation time. We do want to provide a good error message in the event something is missing when they try to `textract` a PDF but don't have the right dependencies installed. That was the intent of `ShellError.is_not_installed` but that clearly isn't working for some command line parsers. 

I have my hands full the next couple of weeks. If someone else wants to take a crack at a PR in the meantime, that would be amazing!

**Comment by deanmalmgren** [comment]
@ad-m @scubedoo187 does installing all of the dependencies fix this problem for you, too? (See @webeng's comment above)

**Comment by ad-m** [comment]
@deanmalmgren , yes, adding missing dependencies fixed that, but for identify dependencies I have modify code to lookup executed command.

**Comment by AusIV** [comment]
@deanmalmgren - The code you linked above  https://github.com/deanmalmgren/textract/blob/master/textract/exceptions.py#L69-L76 was broken by this commit: https://github.com/deanmalmgren/textract/commit/ce7bc0ecde56c12a908aae7e454a0fa03236d911#diff-db5f9029a9aa26a0640244cf8f05910aL81

When they removed the "shell=True" argument, that meant that a missing file would throw an OSError on line 80 of that file, instead of running the pipe and returning an exit code 127. This means that there will never be an exit code 127, and thus it will never hit the ShellError.is_not_installed() condition.

I'm assuming that the "shell=True" argument was removed for security reasons (see the warning here: https://docs.python.org/2/library/subprocess.html#popen-constructor ). Re-adding that argument would be the easiest way to resolve this issue, but we could also restructure the code to check for an OSError when the pipe is created and raise some variation of the ShellError.

**Comment by deanmalmgren** [comment]
@pierre-ernst do you have any suggestions for how to still throw the `exceptions.ShellError` that captures exit code 127 without causing a security issue?

Apparently #114 removed an untested feature that is very convenient for debugging purposes. My bad on not having the test, but I'd appreciate your guidance here before I simply reintroduce `shell=True`, which apparently has some issues (see comment above).

**Comment by pierre-ernst** [comment]
Before the security fix, the 127 exit code was **returned by the shell**.

Maybe catching the OSError Vs ValueError exception will help you determine whether the external program was not found or if it has been called with wrong arguments?

A

**Comment by deanmalmgren** [comment]
@AusIV @pierre-ernst I see. Yes. I like the idea of catching `OSError` and raising some variant of the `ShellError` that displays the `ShellError.not_installed_message`. I might have some airplane time next week where I could work on this, but I'd appreciate a PR if either of you (or anyone else!) are hankering to fix it.

It would also be nice to add some tests that trigger the `ShellError.not_installed_message` so we can be sure to have this functionality going forward as well.

**Comment by AusIV** [comment]
If that's the route you want to go, I'll put together a PR this afternoon!

**Comment by deanmalmgren** [comment]
Excellent; thanks @AusIV !!

**Comment by AusIV** [comment]
PR will probably have to wait until Monday. Production issue popped up this afternoon and ate the time I'd planned to spend on this. In any case, I'll still plan to take this one.

**Comment by deanmalmgren** [comment]
Sure thing. Thanks for the heads up and best of luck with the production issue!

**Comment by deanmalmgren** [comment]
closed by #139
